pkg/kubelet/util/pod_startup_latency_tracker.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/kubelet/metrics"
+	"k8s.io/utils/clock"
+)
+
+// PodStartupLatencyTracker records key moments for startup latency calculation,
+// e.g. image pulling or pod observed running on watch.
+type PodStartupLatencyTracker struct {
+	// protect against concurrent read and write on pods map
+	lock sync.Mutex
+	pods map[types.UID]*perPodState
+	// For testability
+	clock clock.Clock
+}
+
+type perPodState struct {
+	firstStartedPulling time.Time
+	lastFinishedPulling time.Time
+	// first time, when pod status changed into Running
+	observedRunningTime time.Time
+	// log, if pod latency was already Observed
+	metricRecorded bool
+}
+
+func NewPodStartupLatencyTracker() *PodStartupLatencyTracker {
+	return &PodStartupLatencyTracker{
+		pods:  map[types.UID]*perPodState{},
+		clock: clock.RealClock{},
+	}
+}
+
+// ObservedPodOnWatch to be called from somewhere where we look for pods.
+func (p *PodStartupLatencyTracker) ObservedPodOnWatch(pod *v1.Pod, when time.Time) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	// pod Terminated and would never start
```

Comment:
You probably also want to check to see if the pod is Succeeded (which is also a terminal state)

Reply:
Good point. I initially thought, that for a stable cluster we would store all the pods startup data - but you're right, we don't actually need it. Check updated for all terminal states.